### DISCUSSION TITLE
Delete childs with cascade seems not to work with my sqlite DB

### DISCUSSION THREAD

**Discussion by frchalaoux** [sqlalchemy/sqlalchemy#10586] [Usage Questions]
Hi Everyone,

With sqlalchemy 2.x, I have 2 tables, one is the 'riders' table and the other is the 'bikes' table which represents the bikes owned by a rider. If I want to delete a rider I will have to also delete the bikes associated with him. With the code and tables provided, the rider is deleted but the bikes remain in the DB. How can I correct this behavior so that rider's bikes are destroyed if a rider is deleted ? 

For information I use sqlite 3.27.2.

```
def delete_rider_by_id(id: int, db: Session) -> int:
    existing_rider = db.query(RiderDB).filter(RiderDB.rider_id == id)
    if not existing_rider.first():
        return 0
    existing_rider.delete(synchronize_session="auto")
    db.commit()
    return 1
```

```
class RiderDB(Base):
    __tablename__ = 'riders'

    rider_id: Mapped[int] = mapped_column(primary_key=True, index=True)
    lastname: Mapped[str] = mapped_column(String,unique=False,nullable=False)
    firstname: Mapped[str] = mapped_column(String,unique=False,nullable=False)
    nickname: Mapped[str] = mapped_column(String,unique=True,nullable=False)
    email: Mapped[str] = mapped_column(String,nullable=False,unique=True,index=True)

    bikes = relationship("BikeDB", back_populates="owner", cascade="all, delete-orphan")

    owner_id: Mapped[int] =  mapped_column(Integer,ForeignKey("users.user_id"))
    owner = relationship("User", back_populates="riders")
```

```
class BikeDB(Base):
    __tablename__ = 'bikes'

    bike_id: Mapped[int] = mapped_column(Integer, primary_key=True)
    model: Mapped[int] = mapped_column(String(16))

    owner_id: Mapped[int] = mapped_column(Integer, ForeignKey('riders.rider_id'))
    owner = relationship("RiderDB", back_populates="bikes")
```

**Comment by zzzeek** [comment]
hi -

based on the presence of the "synchronize_session" option, I would assume the `.delete()` you are using is some kind of extension  which links `Cls.delete` to the Query -level [bulk delete](https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.delete) feature.  this just emits a DELETE statement and ORM unit of work features like "Cascades" have no ability to take effect.  There should be caveats listed in the linked docs to this effect.

To delete an object within the unit of work, use `session.delete(object)`.  on the next flush , SQLAlchemy will be able to handle those cascade options.

Otherwise, to have all SQL-level DELETES perform this "cascading" on the server, use [ON DELETE CASCADE](https://docs.sqlalchemy.org/en/20/orm/cascades.html#passive-deletes) on your foreign keys, so that any DELETE of any kind of "riders" will also delete the corresponding rows in "bikes".   *however*, the SQL database world has more complications to add here, with SQLite , server level deletes will not work for the Python sqlite driver unless you [follow these steps to allow them to work](https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#foreign-key-support).

**Comment by frchalaoux** [comment]
@zzzeek
 Hi and thanks for your answer :
 
 As I can read from your link [follow these steps to allow them to work](https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#foreign-key-support) I could use : 

 @event.listens_for(Engine, "connect")
def set_sqlite_pragma(dbapi_connection, connection_record):
    cursor = dbapi_connection.cursor()
    cursor.execute("PRAGMA foreign_keys=ON")
    cursor.close()

1/ `connection_record` is not used ?
2/ Concretely in my code how to call this function?
3/ Should I call it for every query ?

**Reply by zzzeek** [reply]
ok lets slow down a bit.  first off, without using any of the event stuff, for the code example you gave, it will work right now if you change it like this:

```py
def delete_rider_by_id(id: int, db: Session) -> int:
    existing_rider = db.query(RiderDB).filter(RiderDB.rider_id == id)
    if not existing_rider.first():
        return 0
    db.delete(existing_rider)
    db.commit()
    return 1

```

that's the form you need to use if you want SQLAlchemy to issue DELETE statements for the "bikes".

Then, beyond that immediate fix, if you also wanted statements like "DELETE * FROM RIDERS" to automatically delete from "bikes" using server side deletion rules, meaning SQLite does the deletion, then you can set up the foreign key constraints to use ON DELETE CASCADE.    However.  Using the Python SQLite driver, we need to associate some additional configuration with the engine.  You are definitely, guaranteed using a SQLAlchemy engine somewhere, however you haven't illustrated where.   If you are using flask-sqlalchemy, then you started out here: https://flask-sqlalchemy.palletsprojects.com/en/3.1.x/quickstart/#configure-the-extension   where they conveniently hide [the engine](https://docs.sqlalchemy.org/en/20/core/engines.html) entirely, so that new users have no idea this very central and important SQLAlchemy object even exists, leading to confusion like this.    So for Flask, the engine is [here](https://flask-sqlalchemy.palletsprojects.com/en/3.1.x/api/#flask_sqlalchemy.SQLAlchemy.engine).

So now that we see where the engine is, we can attach [event listeners](https://docs.sqlalchemy.org/en/20/core/event.html#events) to it.  These are functions that we write and associate with objects, and we can after that forget about them; they take place as the SQLAlchemy object to which they are associated with does its work.  For the Engine, this work involves connecting to databases and emitting commands to the database.   

The specific event hook we are using here is [PoolEvents.connect](https://docs.sqlalchemy.org/en/20/core/events.html#sqlalchemy.events.PoolEvents.connect).   This is a hook that runs whenever SQLAlchemy makes a new connection to the database.  For SQLite we want to run a command on new connections to allow foreign key constraints to work, which is [PRAGMA foreign_keys](https://www.sqlite.org/foreignkeys.html).   SQLite otherwise will not run ON DELETE CASCADE commands if this non-default setting is not configured.   so we implement the hook so that whenever new connections are made, this gets run.

Consider if you have your flask init code somewhere, this all looks like this:

```py
from sqlalchemy import event

app = Flask(__name__)
# configure the SQLite database, relative to the app instance folder
app.config["SQLALCHEMY_DATABASE_URI"] = "sqlite:///project.db"
# initialize the app with the extension
db.init_app(app) 

@event.listens_for(db.engine, "connect")
def set_sqlite_pragma(dbapi_connection, connection_record):
    cursor = dbapi_connection.cursor()
    cursor.execute("PRAGMA foreign_keys=ON")
    cursor.close()

```


Within that event hook, there is no need to worry about "connection_record", we are only concerned with "dbapi_connection", which here is a SQLite connection.

that should be everything, try the easy method first before getting into setting up FOREIGN KEY cascades.

**Reply by frchalaoux** [reply]
Hi zzzeek and a lot for your explanations,

1/ I'm working with fastapi
2/ I try to use your first code 

```
def delete_rider_by_id(id: int, db: Session) -> int:
    existing_rider = db.query(RiderDB).filter(RiderDB.rider_id == id)
    if not existing_rider.first():
        return 0
    db.delete(existing_rider)
    db.commit()
    return 1
```

And I obtained this error : 

```
INFO:     127.0.0.1:54784 - "DELETE /riders/delete-rider/2 HTTP/1.1" 500 Internal Server Error
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 3383, in delete
    state = attributes.instance_state(instance)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'Query' object has no attribute '_sa_instance_state'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/uvicorn/protocols/http/h11_impl.py", line 408, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/uvicorn/middleware/proxy_headers.py", line 84, in __call__
    return await self.app(scope, receive, send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/fastapi/applications.py", line 1115, in __call__
    await super().__call__(scope, receive, send)
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/starlette/applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
    raise exc
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/fastapi/middleware/asyncexitstack.py", line 20, in __call__
    raise e
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/fastapi/middleware/asyncexitstack.py", line 17, in __call__
    await self.app(scope, receive, send)
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/starlette/routing.py", line 718, in __call__
    await route.handle(scope, receive, send)
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/starlette/routing.py", line 276, in handle
    await self.app(scope, receive, send)
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/starlette/routing.py", line 66, in app
    response = await func(request)
               ^^^^^^^^^^^^^^^^^^^
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/fastapi/routing.py", line 274, in app
    raw_response = await run_endpoint_function(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/fastapi/routing.py", line 193, in run_endpoint_function
    return await run_in_threadpool(dependant.call, **values)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/starlette/concurrency.py", line 41, in run_in_threadpool
    return await anyio.to_thread.run_sync(func, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/anyio/to_thread.py", line 33, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 877, in run_sync_in_worker_thread
    return await future
           ^^^^^^^^^^^^
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 807, in run
    result = context.run(func, *args)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/frchalaoux/Documents/Developpement/SMB-all/SMB/src/apis/version1/route_riders.py", line 139, in delete_rider
    delete_rider_by_id(id=id, db=db)
  File "/Users/frchalaoux/Documents/Developpement/SMB-all/SMB/src/db/repository/riders.py", line 134, in delete_rider_by_id
    db.delete(existing_rider)
  File "/Users/frchalaoux/opt/miniconda3/envs/smb/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 3385, in delete
    raise exc.UnmappedInstanceError(instance) from err
sqlalchemy.orm.exc.UnmappedInstanceError: Class 'sqlalchemy.orm.query.Query' is not mapped
```

**Reply by frchalaoux** [reply]
The correct code is :

def delete_rider_by_id(id: int, db: Session) -> int:
    existing_rider = db.query(RiderDB).filter(RiderDB.rider_id == id)
    if not existing_rider.first():
        return 0
    db.delete(existing_rider.**first()**)
    db.commit()
    return 1
    
   Thanks zzzeek.

**Reply by frchalaoux** [reply]
I will try your other proposition next week that seems more touchy to me.
Cheers, and Read you. François-Régis
